exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java

```diff
@@ -0,0 +1,272 @@
+    DistributedLease lease = null;
+    long queryMemory;
+    final DistributedSemaphore distributedSemaphore;
+    try {
+
+      // Only the refresh and queue computation is synchronized.
+
+      synchronized(this) {
+        refreshConfig();
+
+        // get the appropriate semaphore
+        if (cost >= queueThreshold) {
+          distributedSemaphore = clusterCoordinator.getSemaphore("query.large", largeQueueSize);
+          queueName = "large";
+          queryMemory = memoryPerLargeQuery;
+        } else {
+          distributedSemaphore = clusterCoordinator.getSemaphore("query.small", smallQueueSize);
+          queueName = "small";
+          queryMemory = memoryPerSmallQuery;
+        }
+      }
+      logger.debug("Query {} with cost {} placed into the {} queue.",
+                   QueryIdHelper.getQueryId(queryId), cost, queueName);
+
+      lease = distributedSemaphore.acquire(queueTimeout, TimeUnit.MILLISECONDS);
+    } catch (final Exception e) {
+      logger.error("Unable to acquire slot for query " +
+                   QueryIdHelper.getQueryId(queryId), e);
+      throw new QueryQueueException("Unable to acquire slot for query.", e);
+    }
+
+    if (lease == null) {
+      int timeoutSecs = (int) Math.round(queueTimeout/1000.0);
+      logger.warn("Queue timeout: {} after {} seconds.", queueName, timeoutSecs);
+      throw new QueueTimeoutException(queryId, queueName, timeoutSecs);
+    }
+    return new DistributedQueueLease(queryId, queueName, lease, queryMemory);
+  }
+
+  private synchronized void refreshConfig() {
+    long now = System.currentTimeMillis();
+    if (now < refreshTime) {
+      return;
+    }
+    queueThreshold = optionManager.getOption(ExecConstants.QUEUE_THRESHOLD_SIZE);
+    queueTimeout = optionManager.getOption(ExecConstants.QUEUE_TIMEOUT);
+    largeQueueSize = (int) optionManager.getOption(ExecConstants.LARGE_QUEUE_SIZE);
+    smallQueueSize = (int) optionManager.getOption(ExecConstants.SMALL_QUEUE_SIZE);
+    largeToSmallRatio = optionManager.getOption(ExecConstants.QUEUE_MEMORY_RATIO);
+    refreshTime = now + 5000;
+    logger.debug("Configuration: small queue size = {}, large queue size = {}",
+                  smallQueueSize, largeQueueSize);
+    logger.debug("Configuration: cost threshold = {}, timeout = {} ms.",
+                 queueThreshold, queueTimeout);
+    assignMemory();
+  }
+
+  @Override
+  public boolean enabled() { return true; }
+
+  public synchronized ZKQueueInfo getInfo() {
+    refreshConfig();
+    return new ZKQueueInfo(this);
+  }
+
+  private void release(QueueLease lease) {
+    DistributedQueueLease theLease = (DistributedQueueLease) lease;
+    for (;;) {
+      try {
+        theLease.lease.close();
+        theLease.lease = null;
+        break;
+      } catch (final InterruptedException e) {
+        // if we end up here, the loop will try again
+      } catch (final Exception e) {
+        logger.warn("Failure while releasing lease.", e);
```

Comment:
If we can't connect to ZK, then we'll drop out of the ZK-based Drill cluster and the Drillbit will terminate.

To be honest, I don't know why the code is written this way; I picked it up and moved it from the Foreman so it is original code prior to my change. I suppose the idea is that we're holding onto a lease. If we fail to release it, then available query slots will decrease. Do that enough times and no queries will run.

So, seems fair to retry until either 1) the Drillbit exits, or 2) the lease is released.

Probably should add a check of the Drillbit state: have we entered shutdown state?

Once we shutdown, I hope ZK will notice that the lease holder is gone (because the corresponding ZK node is ephemeral) and will force the lease to be released.

Reply:
Good point. Added a mechanism to detect when the Drillbit is shutting down and abandon the loop. However, at present, Drill does not have an idea of a shutdown state. Fortunately, however, DRILL-4286 (graceful shutdown) will provide such a state. We can wire up the mechanisms at that time.